website/docs/error-reference.md

```diff
@@ -1803,6 +1803,63 @@ def foo; [0]; end
 For more information, see
 [Arrays, Hashes, and Generics in the Standard Library](stdlib-generics.md).
 
+## 5027
+
+This error fires when a constant is referenced before it is defined.
+
+```ruby
+puts X # error: `X` referenced before it is defined
+X = 1
+```
+
+```ruby
+module Foo
+  A = X
+    # ^ error: `Foo::X` referenced before it is defined
+  class X; end
+end
+```
+
+Generally, Sorbet is not opinionated about definition-reference ordering. It
+assumes the presence of an autoloader which makes constants available at runtime
+when they are referenced (assuming they are defined somewhere).
+
+However, it is helpful in some basic contexts (like class-loading) to catch this
+issue, regardless of the presence of an autoloader. Please note that this error
+will not fire for all cases. There are some conditions which have to be met for
+the error to fire:
```

Comment:
```suggestion
Generally, Sorbet is not opinionated about definition-reference ordering. It
assumes files are required in the correct order or at the correct times to
ensure that definitions are available before they're referenced.

However, if a constant is defined in a single file, Sorbet can detect when it's
been referenced in that file ahead of its definition (because in the single-file
case, it doesn't matter whether or in what order any require statements
happen). There are some limitations:
```

sorbet does not require an autoloader. rephrasing in terms of require statements.